metadata/src/main/java/org/apache/kafka/image/TopicDelta.java

```diff
@@ -158,11 +160,21 @@ public LocalReplicaChanges localChanges(int brokerId) {
                         new LocalReplicaChanges.PartitionInfo(id(), entry.getValue())
                     );
                     topicIds.putIfAbsent(name(), id());
+                    updateDirectoryIds(brokerId, entry, prevPartition, directoryIds);
                 }
             }
         }
 
-        return new LocalReplicaChanges(deletes, leaders, followers, topicIds);
+        return new LocalReplicaChanges(deletes, leaders, followers, topicIds, directoryIds);
+    }
+
+    private void updateDirectoryIds(int brokerId, Entry<Integer, PartitionRegistration> entry, PartitionRegistration prevPartition, Map<TopicPartition, LocalReplicaChanges.PartitionInfo> directoryIds) {
+        if (prevPartition == null || prevPartition.directory(brokerId) != entry.getValue().directory(brokerId)) {
+            directoryIds.put(
+                    new TopicPartition(name(), entry.getKey()),
+                    new LocalReplicaChanges.PartitionInfo(id(), entry.getValue())
```

Comment:
I think this can just be `entry.getValue().directory(brokerId)`, we don't need the rest

Reply:
updated this.